### ISSUE TITLE
GradientTape host-crash on long raw-tape + engine-ops loop (0.74.0) — distinct path from #279/#283

### ISSUE THREAD

**Issue by ooples** [ooples/AiDotNet.Tensors#312]
## Summary

Even after #283's 10× improvement, **GradientTape still produces host-fatal heap exhaustion** on long-running tests that use raw `using (var tape = new GradientTape<float>())` blocks with engine ops and custom optimizer code (rather than `Transformer<T>.Train`). On Tensors **0.74.0** + AiDotNet 0.166.0, a 192-cell HarmonicEngine test crashed the test host mid-run after ~9 of 192 cells completed (~9,720 Train-equivalent calls). Per-cell `GC.Collect(); GC.WaitForPendingFinalizers(); GC.Collect();` between cells does NOT fully reclaim. Repros distinct from #279/#283 because we don't go through `Transformer.Train` — the leak lives lower in the GradientTape stack itself.

## Why this matters

Research / paper-grade evaluation runs that train many small variants in series (e.g. depth × seed sweeps in a single test process) cannot use `dotnet test` with GradientTape at any nontrivial scale. We've had to:
1. Split sweeps into ever-smaller chunks (32 cells max per process)
2. Force GC between every cell
3. Still hit `Test host process crashed` at L=8 depth + 8 seeds

This blocks the L6 LayerNorm replacement experiments on `ooples/HarmonicEngine` (and would block any paper benchmark battery that needs a few thousand GradientTape rounds per test process).

## Reproduction (minimal, no AiDotNet.Transformer dependency)

This repro uses ONLY `AiDotNet.Tensors` + custom op chain — different code path from #283's `Transformer.Train` repro:

```csharp
using AiDotNet.Tensors.Engines;
using AiDotNet.Tensors.Engines.Autodiff;
using AiDotNet.Tensors.Helpers;
using AiDotNet.Tensors.LinearAlgebra;

const int F = 128, H = 512, V = 256, B = 256, L = 8;
var engine = AiDotNetEngine.Current;
var rng = new Random(42);

// Minimal multi-layer FFN with chained TensorMatMul / TensorMultiply / Tanh / LayerNorm.
var x0 = TensorAllocator.Rent<float>([B, F]);
var targets = TensorAllocator.Rent<float>([B, V]);
var W1s = new Tensor<float>[L]; var b1s = new Tensor<float>[L];
var W2s = new Tensor<float>[L]; var b2s = new Tensor<float>[L];
var gamma = new Tensor<float>[L]; var beta = new Tensor<float>[L];
for (int l = 0; l < L; l++) {
    W1s[l] = TensorAllocator.Rent<float>([F, H]);
    b1s[l] = TensorAllocator.Rent<float>([H]);
    W2s[l] = TensorAllocator.Rent<float>([H, F]);
    b2s[l] = TensorAllocator.Rent<float>([F]);
    gamma[l] = TensorAllocator.Rent<float>([F]); for (int i = 0; i < F; i++) gamma[l].AsWritableSpan()[i] = 1f;
    beta[l] = TensorAllocator.Rent<float>([F]);
}
var Wh = TensorAllocator.Rent<float>([F, V]);
var bh = TensorAllocator.Rent<float>([V]);
var trainable = new List<Tensor<float>>();
foreach (var p in W1s) trainable.Add(p); foreach (var p in b1s) trainable.Add(p);
foreach (var p in W2s) trainable.Add(p); foreach (var p in b2s) trainable.Add(p);
foreach (var p in gamma) trainable.Add(p); foreach (var p in beta) trainable.Add(p);
trainable.Add(Wh); trainable.Add(bh);

// Warmup
for (int i = 0; i < 5; i++) {
    using var tape = new GradientTape<float>();
    var x = (Tensor<float>)x0;
    for (int l = 0; l < L; l++) {
        var hMid = engine.TensorBroadcastAdd(engine.TensorMatMul(x, W1s[l]), b1s[l]);
        var hAct = engine.TensorMultiply(engine.TensorMultiply(hMid, hMid), hMid);
        var Fx = engine.TensorBroadcastAdd(engine.TensorMatMul(hAct, W2s[l]), b2s[l]);
        x = engine.Tanh(engine.TensorAdd(x, Fx));
        x = engine.LayerNorm(x, gamma[l], beta[l], 1e-5, out _, out _);
    }
    var logits = engine.TensorBroadcastAdd(engine.TensorMatMul(x, Wh), bh);
    var grads = tape.ComputeGradients(engine.TensorCrossEntropyLoss(logits, targets), trainable);
    foreach (var g in grads.Values) TensorAllocator.Return(g);
}
GC.Collect(); GC.WaitForPendingFinalizers(); GC.Collect();
long start = GC.GetTotalMemory(true);

// Hot loop — measure retention
const int N = 500;
for (int i = 0; i < N; i++) {
    using var tape = new GradientTape<float>();
    var x = (Tensor<float>)x0;
    for (int l = 0; l < L; l++) {
        var hMid = engine.TensorBroadcastAdd(engine.TensorMatMul(x, W1s[l]), b1s[l]);
        var hAct = engine.TensorMultiply(engine.TensorMultiply(hMid, hMid), hMid);
        var Fx = engine.TensorBroadcastAdd(engine.TensorMatMul(hAct, W2s[l]), b2s[l]);
        x = engine.Tanh(engine.TensorAdd(x, Fx));
        x = engine.LayerNorm(x, gamma[l], beta[l], 1e-5, out _, out _);
    }
    var logits = engine.TensorBroadcastAdd(engine.TensorMatMul(x, Wh), bh);
    var grads = tape.ComputeGradients(engine.TensorCrossEntropyLoss(logits, targets), trainable);
    foreach (var g in grads.Values) TensorAllocator.Return(g);
}
GC.Collect(); GC.WaitForPendingFinalizers(); GC.Collect();
long end = GC.GetTotalMemory(true);
Console.WriteLine($"Retention per call: {(end - start) / (double)N:F0} bytes over {N} calls");
```

Expected behavior: `retention < 10 KB/call` post-Gen2 (#283's stretch-goal). Observed in `ooples/HarmonicEngine` repo with Tensors 0.74.0: heap grows monotonically until host commit limit is exhausted, then `Test host process crashed`.

## Observed crash signature (real from HarmonicEngine L6-E3a)

192 cells of `Phase_L6_E3a_CayleyMobius` (4 depths × 6 variants × 8 seeds) with explicit `GC.Collect()` between cells:
- L=1 batch (48 cells, 6 variants × 8 seeds): completed
- L=2 batch (48 cells): partial — host crashed at cell 51 of 192
- xUnit log: `The active test run was aborted. Reason: Test host process crashed`

Wall-clock at crash: ~30 minutes. Train calls per cell at L=2: ~720. Total Train calls: ~36,000 — same regime as #283's projection table where 0.69.3 retains ~22 GiB at 56K calls. Suggests **the residual leak ratio per call has not improved enough between 0.69.3 and 0.74.0** for our raw-tape + engine-ops use case.

## Hypothesized cause (specific to raw GradientTape + engine ops)

Three places this leak could differ from `Transformer.Train` (already partially fixed in 0.69.x):

1. **Engine-recorded ops**: `engine.TensorMatMul`, `engine.TensorMultiply`, `engine.Tanh`, `engine.LayerNorm`, `engine.TensorAdd`, `engine.TensorSubtract`, `engine.TensorReciprocal`, `engine.TensorBroadcastAdd`, `engine.TensorCrossEntropyLoss` — each may retain a `SavedInputs` clone. `Transformer.Train` paths may bypass some of these via fused ops.

2. **`tape.ComputeGradients` output dictionary**: in our pattern we `foreach (var g in grads.Values) TensorAllocator.Return(g);` but the returned dictionary itself, plus any tape-side reverse-mode metadata, may persist after `using` scope ends.

3. **`engine.LayerNorm`'s out-params** (`out _, out _`): we discard the cached `mean`/`var` outputs. If the tape captures them as `SavedInputs` for a backward step, they may leak even with the `out _` discard.

## Suggested triage

1. Run the above repro on Tensors HEAD (0.75.0+) and report retention/call.
2. If still > 10 KB/call: heap-diff between Train call N and N+1 with the raw-tape repro (different rooting pattern than #283's Transformer.Train repro).
3. Add a CI canary that does 1000 raw-tape ComputeGradients calls and asserts retention < threshold. (#283 also recommended this; it's apparently still missing.)

## Cross-references

- AiDotNet.Tensors#279, #283 — predecessors. Closed, but consumer-facing crash recurs on 0.74.0 in raw-tape use.
- ooples/AiDotNet#1227 — consumer-side report.
- ooples/HarmonicEngine `theorem/l6-spectral-regenerator` branch, `tests/Learning/Phase_L6_E3a_CayleyMobius.cs` — full real-world reproducer.

## Repro environment

- `AiDotNet.Tensors` **0.74.0**
- `AiDotNet` 0.166.0
- .NET 10.0.7
- Windows 11 Pro, 16-core
